### DISCUSSION TITLE
Using model/column to execute filtering in multiple threads

### DISCUSSION THREAD

**Discussion by youtian001** [sqlalchemy/sqlalchemy#7409] [Usage Questions] [answered]
Hi Team,

I see a problem that when I use the model/column to execute filtering in multiple threads, just the first query works and all the next queries just get 0 result.

My query is like:

`
session.query(MyTable).filter(MyTable.id == id).all()
`

I found this post and mentioned a potential problem of the filter:
https://stackoverflow.com/questions/48128705/sqlalchemy-scoped-session-is-not-getting-latest-data-from-db

With the solution, replacing `MyTable.id == id` to `text(f"id = {id}")`, the queries works fine.

But the root cause is not mentioned in the post. I'm not sure if it's a known issue of SQLAlchemy or it's caused by wrong usage of ORM+filter?

Thanks.

**Comment by zzzeek** [comment]
hey there -

there is absolutely no reason whatsoever that using an ORM expression within filter() would have any interaction of any kind with the fact that threads are in use, unless one is actively modifying the structure of the `MyTable` class at runtime which would be very poor practice (and also fail a lot more badly).

The stackoverflow issue linked is three years old so this also eliminates the possibility that this person was having an issue related to SQL caching, however their conclusion that the problem is the filter() method is likely not what was actually happening.

that you appear to be observing the same thing is pretty important and if you were able to provide lots more details here, that would be helpful to the SQLAlchemy project so that we may know what pattern is in use here, involving log output as well as complete models / runnable sample code / sqlalchemy version in use.

I think in your case you would need to look at [SQL echoing](https://docs.sqlalchemy.org/en/14/core/engines.html#more-on-the-echo-flag)  to see what SQL is being emitted for filter() when you are using the two different forms of expression so that the difference may be noted.

**Reply by CaselIT** [reply]
> With the solution, replacing MyTable.id == id to text(f"id = {id}"), the queries works fine.

also don't do that if `id` is not trusted (and even if it is) since you are effectively opening yourself to a sql injection

**Comment by youtian001** [comment, marked as answer]
Thanks folks for the enlightening response!

I turn on the SQL echoing and find out the root cause which is in my code and not SQLAlchemy related.

A mistake was made to give parameter of a method a mutable value as the default and the default value is changed in the method. It's like:

```
def query(id, filters=[]):
    filters.append(MyTable.id == id)
    ...
```

So the second time using this method, the filters have previously appended condition.

**Reply by CaselIT** [reply]
Ah that's a classic!, I think sooner or later everyone gets bitten by that.
